src/libostree/ostree-sysroot-deploy.c

```diff
@@ -2127,147 +2211,73 @@ ostree_sysroot_write_deployments_with_options (OstreeSysroot     *self,
-       * But this is a "nice-to-have" for the bootloader UI, so failure
-       * here is not fatal to the whole operation.  We just gracefully
-       * fall back to the deployment index. */
-      (void) ostree_sysroot_get_repo (self, &repo, cancellable, NULL);
-
-      /* Only show the osname in bootloader titles if there are multiple
-       * osname's among the new deployments.  Check for that here. */
-      for (guint i = 1; i < new_deployments->len; i++)
-        {
-          const char *osname_0 = ostree_deployment_get_osname (new_deployments->pdata[0]);
-          const char *osname_i = ostree_deployment_get_osname (new_deployments->pdata[i]);
-          if (!g_str_equal (osname_0, osname_i))
-            {
-              show_osname = TRUE;
-              break;
-            }
-        }
-
-      for (guint i = 0; i < new_deployments->len; i++)
-        {
-          OstreeDeployment *deployment = new_deployments->pdata[i];
-          if (!install_deployment_kernel (self, repo, new_bootversion,
-                                          deployment, new_deployments->len,
-                                          show_osname, cancellable, error))
-            goto out;
-        }
-
-      /* Create and swap bootlinks for *new* version */
-      if (!create_new_bootlinks (self, new_bootversion,
-                                 new_deployments,
-                                 cancellable, error))
-        goto out;
-      if (!swap_bootlinks (self, new_bootversion, new_deployments,
-                           cancellable, error))
-        goto out;
-
-      g_debug ("Using bootloader: %s", bootloader ?
-               g_type_name (G_TYPE_FROM_INSTANCE (bootloader)) : "(none)");
-
-      if (bootloader)
-        bootloader_is_atomic = _ostree_bootloader_is_atomic (bootloader);
+        return FALSE;
+      bootloader_is_atomic = bootloader != NULL && _ostree_bootloader_is_atomic (bootloader);
 
-      if (bootloader)
+      /* Note equivalent of try/finally here */
+      gboolean success = write_deployments_bootswap (self, new_deployments, opts, bootloader,
+                                                     &syncstats, cancellable, error);
+      /* Below here don't set GError until the if (!success) check */
+      if (boot_was_ro_mount)
         {
-          if (!_ostree_bootloader_write_config (bootloader, new_bootversion,
-                                                cancellable, error))
+          if (mount ("/boot", "/boot", NULL, MS_REMOUNT | MS_RDONLY | MS_SILENT, NULL) < 0)
             {
-              g_prefix_error (error, "Bootloader write config: ");
-              goto out;
+              /* Only make this a warning because we don't want to
+               * completely bomb out if some other process happened to
+               * jump in and open a file there.  See above TODO
+               * around doing this in a new mount namespace.
+               */
+              int errsv = errno;
```

Comment:
I know this isn't new, but that line confuses me. Was the intent here to save `errno` *before* the "finally" `mount` and restore it afterwards?

Reply:
I think it's more that in the past I've been bitten by the error handling functions themselves overwriting `errno`.  Imagine if e.g. you want to do `g_strdup_printf()` to format an error message, and then go on to reference `errno`..  But in this case it's safe to drop.